keps/sig-network/2112-netpol-namespace-name-selector/README.md

```diff
@@ -0,0 +1,393 @@
+retroactively added to namespaces pretty easily, even if your namespace isn't 
+intended to be able to send traffic somewhere: 
+
+```
+we need to clarify how namespaces are matched. I'm pretty 
+uncomfortable with using labels here, since anyone can add labels at 
+any time to any namespace and thus send traffic to your pod if they 
+know the label you're using. If it were simply a namespace name, it's 
+much easier to specifically allow only namespaces you control since you 
+know your own namespace names. 
+```
+
+- Another recent argument was made on the basis of targeting namespaces 
+using names, rather then labels, on the basis of **sheer convenience**.
+
+```
+While matching things like pods etc by label is certainly worthwhile, when 
+matching a namespace I suspect the majority of the time you only want to match 
+a single namespace. It been great match against the name rather than just a 
+label. I suspect most people don't think to add labels to the namespaces.
+```
+
+The latter comment received 9 likes as a github issue - indiciating the 
+general popularity of this as a request, and that correlates well to feedback 
+we've seen in the broader community as well.
+
+We thus conclude that, even though targeting an object using its name is not 
+normal in K8s, in the case of NetworkPolicies, the overwhelming need for 
+universal, easy to define security boundaries, makes a strong case for 
+amending the API to support a "special" selector for namespace names that is 
+**independent** of labels.
+
+### Goals
+
+- Add a `namespaceNames` option (which is additive to the current `matchLabels` selector).
+- Allow multiple `namespaceNames` values, so that many namespaces can be 
+allowed using the same field 
+
+### Non-Goals
+
+- Support matching of wildcard namespaces
+- Changing the internal details of how namespaceSelector works with arbitrary syntax
+- Depending on more sophisticated features (like virtual labels) to accomplish 
+this feature without changing the API (note that, if someone were 
+to write a virtualLabels KEP, however, it might change the trajectory of this 
+KEP, which is worth discussing)
+
+### User Stories (Optional)
+
+See motivation section for user stories.  This KEP collects concrete data and opinions from several individuals over the past few years, hence we do not include generic user stories.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+- NetworkPolicy providers may **opt-out**, initially to support this construct, and dilligent communication with CNI providers will be needed to make sure its widely adopted, thus, this feature needs to be backwards compatible with the existing v1 api.
+- We thus need to make sure that hidden defaults don't break the meaning of existing policys, for example:
+  - if `namespaceNames` field is retrieved by an older client which **doesn't** yet support it:
```

Comment:
This is a huge liability.

If you have policy that says:

```
from:
    namespaceNames:
        - foo
    podSelector:
        app: myapp
```

That explicitly means to allow traffic from namespace "foo" AND NOT from *this* namespace.  If you have an old client who doesn't know about the new field you will deserialize as:

```
from:
    podSelector:
        app: myapp
```

That explicitly means to allow traffic from *this* namespace.  That's a VERY different meaning.

I'm not sure what the best way to mitigate this is.  This is a general API evolution problem.  We're adding a new option to a one-of and we need to accommodate clients that can't see it.  In this case, we really want to fail-closed.

As proposed, this API doesn't fail-closed.

One option is to add a new sub-struct that contains BOTH the list of namespace names and the pod selector.  This is redundant BUT it dodges the failure mode.  Now the above example becomes:

 ```
from:
    namespaceList:
        namespaceNames:
          - foo
        podSelector:
          app: myapp
```

Which the old clients read as:

```
from: {}
```

I am very open to alternate ideas, but I think "old client fails closed" is probably a requirement.

Reply:
curiously, virtual labels do not suffer this failure mode...

Reply:
Ahhhh so, now that we're adding a new way to select things, if an old thing ignores that selector, a new policy will appear to be extremely promiscuous, because that new selector will be perceived as nil.

Reply:
by virtual label, do we roughly mean "put junk in the selector that always mismatches unless a client is aware of the secret syntax" ?  i.e. 

```       
    namespaceSelector:
        ^virtualNamespaceName: "foo"
```

if so , yeah your right thats pretty clean .

Reply:
Just to make sure we get to the bottom of the assumptions here (strongly leaning towards synth. labels...)... but to complete this thought of your workaround... 

2.  Considering your workaround idea...  Seems like an arbitrary distinction just making sure im correct, by `from: {}` you really mean

```
from: []
```

right?  from is a list of PeerPolicies, not a map.

... In any case, that `from: []` thingy... thats a `DENY ALL`... right ? (Theres no peers, so theres nothing to match an allow against).    
- in the current default, your allowing more traffic then you intended to
- in the deny all scenario, you are denying more traffic then you intended to 

so, not sure why one is inherently better then the other.

Reply:
# 3  ... Synthetic labels... same pitfall, just a different default  ?

Ok so, after thinking about (1) and (2) above, i came to the conclusion that, synthetic labels are just "different" not "better".... unless theres a hidden premise im missing.

If a client isnt capable of interprettying a synthetic match label for a namespace, then it will always fail matching the namespace, and thus it will ALSO have default "Deny All" semantics.  So Im assuming "deny all is the right behaviour when you dont understand a field" is what your going for.  

.... but of course LMK if im missin somethin, im not the best w/ this API stuff :)

Reply:
> by virtual label, do we roughly mean "put junk in the selector that always mismatches unless a client is aware of the secret syntax"

Caveat: making it up as I go.  No secret syntax.  We own the
`*.kubernetes.io/*` prefix, so you could imagine defining that
`virtual.kubernetes.io/name` is always set to the resource's name.  Now, I
don't know what that really means implementation wise - do we actually store
that or populate it on read or somnething else.  I don't know if apimachinery
would go for it.  I don't know what other issues it would present (e.g. scale).

For sanity, we could say that any attempt to write such a label either fails or
is just discarded.

> QUESTION: just making sure im correct, by from: {} did you really mean

From API docs for `From []NetworkPolicyPeer`:  

"""
If this field is empty or missing, this rule matches all sources (traffic not
restricted by source). If this field is present and contains at least one item,
this rule allows traffic only if the traffic matches at least one item in the
from list.
"""

So the result is a list of one element, and that element having no fields set
falls into the second part: "...allows traffic only if the traffic matches..."

> that `from: []` thingy... thats a DENY ALL... right ? 

NOTE: `from: []` means "from all".  `from: [{}]` means deny all.  I specified
`from: {}` which is wrong - It should have been `[{}]`.  The list HAS an
element, but the client doesn't know what it said, so it is morally `{}`.

> ts still quite far from the original intent of allowing specific traffic from a ns/pod combination, albeit, its more restrictive rather then less.

I expect fail-closed will make people less angry than the alternative :)

> synthetic labels are just "different" not "better".... unless theres a hidden premise im missing. That is, they still result in a wrong interpretation of a rule, because, like any rule, you cannot follow it if you dont understand the grammer.

That's why no magic syntax.  Think of it as a label that every resource has
without specifying.

Reply:
So following what we've discussed in our meeting today, is that possible that we create a validation and/or a mutation in the api side? Like:

* If ``namespaceName > 0`` , then verify if ``namespaceSelector > 0``
* If ``namespaceSelector < 1``, then mutate the object and create a ``namespaceSelector`` that will never be selected

So one way to do this would be to create a [well known label](https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/) like networking.kubernetes.io/null-namespace and in the situation above, try to select this label.

Another solution could be to select namespace using a [matchExpression](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#labelselector-v1-meta) with a never targeted expression. One could make a namespaceSelector with a labelSelector and not the same labelExpression, because per the doc:
```
A label selector is a label query over a set of resources. The result of matchLabels and matchExpressions are ANDed. An empty label selector matches all objects. A null label selector matches no objects.
```

So the following will always select nothing (unless my logics are wrong, and the ANDed in the description above is something like "the union of the results of both selectors", and not "if x && y")

```
namespaceSelector:
  matchLabels:
      noOp: true
  matchExpression:
      - {key: noOp, operator: DoesNotExist}
```
Because if someone creates a noOp label, the second statement will be false, so null selector. If there's no label, the first is false again. But again, my logic can be reaaallly wrong :)


Other way that @jayunit100  suggested would be to insert anything random to be selected.

Reply:
I am not sure what you're getting at here - maybe you can explain a bit more?

> * If ``namespaceName > 0`` , then verify if ``namespaceSelector > 0``
> * If ``namespaceSelector < 1``, then mutate the object and create a ``namespaceSelector`` that will never be selected

What is a valid selector that can never select anything?

> So one way to do this would be to create a [well known label](https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/) like networking.kubernetes.io/null-namespace and in the situation above, mutate the network policy and try to select this label for the namespace.

Two problems.  First, we generally do not modify user input in surprising ways
(and I think this might qualify :).  Second, this means we have to define what
it means when both namespaceNames and namespaceSelector are defined at the same
time.  I am not very excited about carving out a special case here.  E.g.

If namespaceSelector is specifed AND it has one value AND that value is the
"magic word", then don't use it and instead do this other thing.

> Another solution could be to select namespace using a [matchExpression](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#labelselector-v1-meta) with a never targeted expression. One could make a namespaceSelector with a labelSelector and not the same labelExpression, because per the doc:
> ```
> A label selector is a label query over a set of resources. The result of matchLabels
> and matchExpressions are ANDed. An empty label selector matches all objects.
> A null label selector matches no objects.
> ```
>
> So the following will always select nothing (unless my logics are wrong, and the ANDed in the description above is something like "the union of the results of both selectors", and not "if x && y")
>
> ```
> namespaceSelector:
>   matchLabels:
>       noOp: true
>   matchExpression:
>       - {key: noOp, operator: DoesNotExist}
> ```
> Because if someone creates a noOp label, the second statement will be false, so null selector. If there's no label, the first is false again. But again, my logic can be reaaallly wrong :)

That's kind of a hack.  It's impressive in a way, but a hack nonetheless.  See
my above statement about magic words and special-cases.

> Other way that @jayunit100  suggested would be to insert anything random to be selected.

I wasn't there, so I don't know for sure what this means, but it feels like
we're thrashing around looking for a clever hack.  Clever is the enemy of
obvious.  I'd rather just evolve this toward something un-clever.

E.g. We have:

```
PodSelector *metav1.LabelSelector
NamespaceSelector *metav1.LabelSelector
IPBlock *IPBlock
```

We could evolve to something like:

```
// Deprecated - use Pods.Selector
PodSelector *metav1.LabelSelector
// Deprecated - use Namespaces.Selector or Pods.Namespaces.Selector
NamespaceSelector *metav1.LabelSelector

// One of the following may be set.  Consumers should check them in the same
// order they are defined here.  Over time, new options may be added.  Clients
// which do not understand the new options should fail closed.
Pods *PodsPeer
Namespaces *NamespacesPeer
IPBlock *IPBlock
```

and define:

```
type PodsPeer struct {
    // Must be non-empty
    Selector *metav1.LabelSelector
    // Optional, ANDed with selector
    Namespaces *NamespacesPeer
}

type NamespacesPeer struct {
    // One of the following may be set.
    Selector *metav1.LabelSelector
    Names []string
}
```

This is just hand-waving, but hopefully you can see where I am trying to go
with it.

Reply:
Yeah I actually agree that the whole default label / noop label / timestamp label is all a hack.  It's like, even if we do it, it's like making the API worse as a whole.

**If** perfectly non surprising  backward compatibility with other v1 CNIs is a must, then This KEP feels increasingly to me like it should be either an API change (v2) 
, a new type of selector, or a potential "default/synthesized" labels, or "netpol operator" knob. 

That *if* above is maybe what I should bring up at the next sig net call :).

Reply:
> What is a valid selector that can never select anything?

Yes, sorry, I was writing in a rush. The statement here is "if I have inserted namespace by name selector, then I should verify if namespace by label is empty. If it's empty, then I should insert anything that turns it not empty, but never matched. This way if the CNI does not know about the namespace by name selector, it will still not select anything because it knows a namespace by label that doesn't select nothing -> default deny"

> First, we generally do not modify user input in surprising ways (and I think this might qualify :)

Indeed, this was a doubt we had during the meeting, if this was plausible or not.

> Second, this means we have to define what it means when both namespaceNames and namespaceSelector are defined at the same time.

IMO this should be an "OR", because when you specify a namespaceName you already know it's name, so it doesn't make sense to select also it's label. The namespaceSelector is for when you don't have specific namespace information (or want to select a group of namespaces).

> That's kind of a hack. It's impressive in a way, but a hack nonetheless. 

I don't know if I thank and feel 'proud' or with some sort of shame. We call this in Portuguese a "Gambiarra", but I need more time to explain the real meaning. It can be summarized as "a nasty ugly hack that no one knows how it works, why it works, but somehow it works". I agree, I don't like this suggestion, but as this was in my head as a possibility I had to write at least so we could discard it

> This is just hand-waving, but hopefully you can see where I am trying to go with it.

Yes, I got it. But going this way, are we saying that we're deprecating fields in a stable API (Netpol is v1)? If so then Jay is right in his statement and we should go to the way of a v2 api, right?

Reply:
Oh also I've remembered, this sort of case makes more and more concrete the needs and concerns that @danwinship has raised in this [KEP](https://github.com/kubernetes/enhancements/pull/2137).

@thockin I would suggest, if you have some bandwidth to take a look into Dan's Kep before coming back to this  (and maybe to the PortRanges) because it tries to solve this sort of problems in a way I think might be a concern not only for CNIs, but also for CSIs and other stuffs that might plug into Kubernetes using stable APIs :)

Reply:
So #2137 would solve this if everyone already implemented #2137...

One way to do this would be to do "NetworkPolicy v2" **in the Kubernetes sense of v2, not in the "NetworkPolicy++" sense**. We add a new `networkingv2.NetworkPolicy` that is a slightly-cleaned-up version of `networkingv1.NetworkPolicy`, and we add `minVersion` to both `networkingv1.NetworkPolicySpec` and `networkingv2.NetworkPolicySpec`. Then in the future when we want to add new features, we can add them in a nice way to `networkingv2.NetworkPolicy` and in an ugly-but-backward-compatible way to `networkingv1.NetworkPolicy` (with the expectation being that people wouldn't actually try to use the new features via the v1 API so it doesn't matter if it's ugly and hard to write).

eg, the user creates a policy:

```
# NB: "v2"
apiVersion: networking.k8s.io/v2
kind: NetworkPolicy
spec:
  podSelector:
    app: myServer
  ingress:
    - from:
        # Allow connections from all pods in namespaces labeled "special: true"
        # and from pods labeled "app: myapp" in namespace "foo"
        - namespaceSelector:
            matchLabels:
              special: true
        - namespaceNames:
            - foo
          podSelector:
            app: myapp
```

per #2137, the apiserver sees `namespaceNames` and so adds `spec.minVersion: 1.21` at creation time.

A network plugin that implements `networkingv2.NetworkPolicy` and supports the version 1.21 features can then read the policy and interpret it correctly.

A network plugin that implements `networkingv2.NetworkPolicy` but only supports 1.20 and earlier would see the `minVersion`, know that it can't implement this policy, and set appropriate `status` conditions to inform the user.

An old network plugin that doesn't implement `networkingv2.NetworkPolicy` would see the new NetworkPolicy via the `networkingv1` API, which would look something like:

```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
spec:
  minVersion: 1.21
  podSelector:
    app: myServer
  ingress:
    - from:
        - namespaceSelector:
            matchLabels:
              special: true
    - fromV2:
        - namespaceNames:
            - foo
          podSelector:
            app: myapp
          v2Index: 1
```

which, after discarding the fields it doesn't understand is:

```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
spec:
  podSelector:
    app: myServer
  ingress:
    - from:
        - namespaceSelector:
            matchLabels:
              special: true
```

so it would still see the `namespaceSelector` rule that it understands, but it wouldn't see the peer containing the `namespaceNames` rule at all. (So it would end up implementing a more restrictive policy rather than a less restrictive one. Of course we could define the v2↔v1 mapping any way we wanted if we wanted some different behavior here.)

Reply:
> Of course we could define the v2<->v1 mapping any way we wanted if we wanted some different behavior here.

actually it would be a lot simpler to just say that if there were any "new" fields anywhere in `ingress`, then the entire `ingress` section gets moved to `ingressV2` in the v1 version, so v1 clients would just see it as a policy that isolates some pods but doesn't allow any access to them. Likewise with `egress`.

Reply:
Yup, we need to solve this by some kind of indirect means, tldr.

Reply:
Hiya dan quick q as I attempt to formalize this discussion:

Can you clarify what you mean by 


"it as a policy that isolates some pods but doesn't allow any access to them"....  the "but" is making my brain spin  there, since isolating poss and not allowing access to pods are potentially the same operation, not opposed in any obvious way.

Reply:
Like, normally a NetworkPolicy says "isolate these pods (implicitly) and then allow access from X, Y, and Z". If you "hide" `spec.ingress` then the policy ends up saying "isolate these pods, and that's it"

Reply:
> If perfectly non surprising backward compatibility with other v1 CNIs is a must

It's unlikely that there is a perfect solution here.  Only "acceptable", which
(IMO) means "fail closed".

Virtual labels is the closest to ideal FOR THIS SPECIFIC CASE, but it's a bit
of a detour to the objective here.

> are we saying that we're deprecating fields in a stable API

Deprecating fields is allowed, just not REMOVING them.  A v2 API would allow us
to clean up the API a little, but not really much.  We still need fidelity in
round-trip.  There's no panacea.

> per #2137, the apiserver sees namespaceNames and so adds spec.minVersion: 1.21 at creation time.

I'll go read that KEP, but I will say that micro-versioning of APIs has been
discussed, but it comes with its own set of non-trivial costs.  As you
describe, SOMEONE needs to taste the incoming objects and set the API versions,
including across update operations.  Absent a general mechanism to do this, I
have little faith it will work.

Or we make the user set it, and deal with all sorts of error handling when they
get it wrong.

> so it would still see the namespaceSelector rule that it understands, but it wouldn't see the peer containing the namespaceNames rule at all. (So it would end up implementing a more restrictive policy rather than a less restrictive one

We don't need micro-versioning to achieve this, though, just thoughful API
design?

Reply:
Agree lack of a way to match on name is a friction point.  The Calico team have some experience here; when we convert a namespace to our datamodel, we add a label in our namespace so that the user can match on the name.  That seems to work fine for our users.

There's plenty of precedent in the k8s API for auto-added labels (such as all the ones on the Node resource).  Seems like the established norm is that those are just added to the datamodel as "real" labels that can be read as normal with kubectl.  If the label was added as a "real" label like that, support would just fall out in all the CNI plugins and users would be able to do more advanced matches where they want to do so.

Reply:
Node objects are more-or-less owned by Kubernetes itself though, so the fact that Kubernetes adds labels to Nodes isn't an example of "Kubernetes adding labels to other people's objects", it's just a slightly weird case of "owners adding labels to their own objects".

Reply:
> So #2137 would solve this if everyone already implemented #2137...

after some discussion there, #2137 solves the problem of "the apiserver knows about this feature but the network plugin doesn't" (user tries to apply a policy using this feature to a 1.22 cluster running a 1.20-era network plugin) but it doesn't solve the problem when the apiserver doesn't know about the feature (user tries to apply a policy using this feature to a 1.20 cluster) unless the user manually specifies `minVersion` on the policy, which is ugly.